pipeline/execout/cachev1/cache.go

```diff
@@ -0,0 +1,329 @@
+		return false, fmt.Errorf("loading cache: %w", err)
+	}
+	return found, nil
+
+}
+func (c *OutputCache) Load(ctx context.Context, blockRange *block.Range) error {
+	c.logger.Debug("loading cache", zap.String("module_name", c.ModuleName), zap.Object("range", blockRange))
+	c.kv = make(outputKV)
+
+	filename := ComputeDBinFilename(blockRange.StartBlock, blockRange.ExclusiveEndBlock)
+	c.logger.Debug("loading outputs data", zap.String("file_name", filename), zap.String("cache_module_name", c.ModuleName), zap.Object("block_range", blockRange))
+
+	err := derr.RetryContext(ctx, 3, func(ctx context.Context) error {
+		objectReader, err := c.Store.OpenObject(ctx, filename)
+		if err != nil {
+			return fmt.Errorf("loading block reader %s: %w", filename, err)
+		}
+
+		if err = json.NewDecoder(objectReader).Decode(&c.kv); err != nil {
+			return fmt.Errorf("json decoding file %s: %w", filename, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("retried: %w", err)
+	}
+
+	c.CurrentBlockRange = blockRange
+	c.logger.Debug("outputs data loaded", zap.String("module_name", c.ModuleName), zap.Int("output_count", len(c.kv)), zap.Stringer("block_range", c.CurrentBlockRange))
+	return nil
+}
+
+func (c *OutputCache) save(ctx context.Context, filename string) error {
+	c.logger.Info("saving cache", zap.String("module_name", c.ModuleName), zap.Stringer("block_range", c.CurrentBlockRange), zap.String("filename", filename))
+
+	buffer := bytes.NewBuffer(nil)
+	err := json.NewEncoder(buffer).Encode(c.kv)
+	if err != nil {
+		return fmt.Errorf("json encoding outputs: %w", err)
+	}
+	cnt := buffer.Bytes()
+
+	go func() {
+		err = derr.RetryContext(ctx, 3, func(ctx context.Context) error {
+			reader := bytes.NewReader(cnt)
+			return c.Store.WriteObject(ctx, filename, reader)
+		})
+		if err != nil {
+			c.logger.Warn("failed writing output cache", zap.Error(err))
+		}
+	}()
+
+	return nil
+}
+
+func (c *OutputCache) String() string {
+	return c.Store.ObjectURL("")
+}
+
+func (c *OutputCache) ListContinuousCacheRanges(ctx context.Context, from uint64) (block.Ranges, error) {
+	cachedRanges, err := c.ListCacheRanges(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("listing cached ranges %q: %w", c.ModuleName, err)
+	}
+	out := listContinuousCacheRanges(cachedRanges, from)
+	return out, nil
+}
+
+func (c *OutputCache) ListCacheRanges(ctx context.Context) (block.Ranges, error) {
+	var out block.Ranges
+	err := derr.RetryContext(ctx, 3, func(ctx context.Context) error {
+		if err := c.Store.Walk(ctx, "", func(filename string) (err error) {
+			r, err := fileNameToRange(filename)
+			if err != nil {
+				return fmt.Errorf("getting range from filename: %w", err)
+			}
+			out = append(out, r)
+			return nil
+		}); err != nil {
+			return fmt.Errorf("walking cache ouputs: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	sort.Slice(out, func(i, j int) bool {
+		return out[i].StartBlock < out[j].StartBlock
+	})
+
```

Comment:
You can sort.Sort() directly the `block.Ranges` which has everything needed already, doesn't it? If need to be reversed, there's `sort.Reversed()` or something.

Reply:
https://github.com/streamingfast/substreams/blob/bob/block/range.go#L119-L121